akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala

```diff
@@ -0,0 +1,534 @@
+      val (prefixF, suffixF) = Source
+        .empty[Int]
+        .flatMapPrefixMat(0) { prefix =>
+          prefix should be(empty)
+          Flow[Int].mapMaterializedValue(_ => prefix)
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      prefixF.futureValue should be(empty)
+      suffixF.futureValue should be(empty)
+    }
+
+    "propagate errors during flow's creation when n = 0" in assertAllStagesStopped {
+      val (prefixF, suffixF) = src10()
+        .flatMapPrefixMat(0) { prefix =>
+          prefix should be(empty)
+          throw TE("not this time my friend!")
+          Flow[Int].mapMaterializedValue(_ => prefix)
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      prefixF.failed.futureValue should be(a[NeverMaterializedException])
+      prefixF.failed.futureValue.getCause === (TE("not this time my friend!"))
+      suffixF.failed.futureValue should ===(TE("not this time my friend!"))
+    }
+
+    "propagate materialization failures when n = 0" in assertAllStagesStopped {
+      val (prefixF, suffixF) = src10()
+        .flatMapPrefixMat(0) { prefix =>
+          prefix should be(empty)
+          Flow[Int].mapMaterializedValue(_ => throw TE("Bang! no materialization this time"))
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      prefixF.failed.futureValue should be(a[NeverMaterializedException])
+      prefixF.failed.futureValue.getCause === (TE("Bang! no materialization this time"))
+      suffixF.failed.futureValue should ===(TE("Bang! no materialization this time"))
+    }
+
+    "run a detached flow" in assertAllStagesStopped {
+      val publisher = TestPublisher.manualProbe[Int]()
+      val subscriber = TestSubscriber.manualProbe[String]()
+
+      val detachedFlow = Flow
+        .fromSinkAndSource(Sink.cancelled[Int], Source(List("a", "b", "c")))
+        .map { x =>
+          println(x)
+          x
+        }
+        .via {
+          Flow.fromSinkAndSource(Sink.fromSubscriber(subscriber), Source.empty[Int])
+        }
+      val fHeadOpt = Source
+        .fromPublisher(publisher)
+        .flatMapPrefix(2) { prefix =>
+          prefix should ===(0 until 2)
+          detachedFlow
+        }
+        .runWith(Sink.headOption)
+
+      subscriber.expectNoMessage()
+      val subsc = publisher.expectSubscription()
+      subsc.expectRequest() should be >= 2L
+      subsc.sendNext(0)
+      subscriber.expectNoMessage()
+      subsc.sendNext(1)
+      val sinkSubscription = subscriber.expectSubscription()
+      //this indicates
+      fHeadOpt.futureValue should be(empty)
+
+      //materializef flow immediately cancels upstream
+      subsc.expectCancellation()
+      //at this point both ends of the 'external' fow are closed
+
+      sinkSubscription.request(10)
+      subscriber.expectNext("a", "b", "c")
+      subscriber.expectComplete()
+    }
+
+  }
+
+}
```

Comment:
A bunch of uncommented statement across this file, uncomment or remove

Reply:
if you're happy, I'm happy 😎